lib/srv/app/server.go

```diff
@@ -542,15 +552,33 @@ func (s *Server) ForceHeartbeat() error {
 }
 
 // HandleConnection takes a connection and wraps it in a listener so it can
-// be passed to http.Serve to process as a HTTP request.
+// be passed to http.Serve to process as a HTTP request. This method blocks
+// until the connection closes.
 func (s *Server) HandleConnection(conn net.Conn) {
-	// Wrap the listener in a TLS authorizing listener.
-	listener := newListener(s.closeContext, conn)
-	tlsListener := tls.NewListener(listener, s.tlsConfig)
-
-	if err := s.httpServer.Serve(tlsListener); err != nil {
+	// Wrap the TLS authorizing conn in a single-use listener.
+	tlsConn := tls.Server(conn, s.tlsConfig)
+	listener := newListener(s.closeContext, tlsConn)
+
+	// build a channel that will close when tlsConn is closed
+	closeCh := s.httpConnState.Channel(tlsConn, http.StateClosed)
```

Comment:
I am curious, can this miss an event if the connection closes before this channel is returned, and what would happen in that case?

Reply:
No state events will be missed. Passing a closed connection to Serve will emit New and Closed events. But it is important to create the channel before calling Serve, as outlined in godoc, so callers don't miss the closed state.

I thought about tracking the last state and emitting it if Channel is called after a state has already been emitted, but decided against because it doesn't allocate any resources to manage the subscription until Channel is called. The only memory allocation for connections without a subscription is a map key and nil value. The trade off is whether we want to allocate memory to track the last state for all connections regardless of subscription or require the caller to call Channel before Serve. I could argue either way. It isn't a lot of memory.

The typical scenario, when not calling serve with a fake listener, requires extracting the connection from the http request context in an http handler. Since the handler was called, the connection is probably still open and emitting events. But maybe not?

LMK what you think.